### ISSUE TITLE
Create `test-sbt` project for SBT integration with ZIO Test

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#1259]
In a new `test-sbt` project, which depends on `test` project, we will provide integration with SBT, by providing an implementation of `TestFramework`.

This can be built off @NeQuissimus's work in [this pull request](https://github.com/zio/zio/pull/723/files). 

Using reflection, we should identify the set of all `RunnableSpec`, and then run them as per SBT integration requirements.

**Comment by dkarlinsky** [comment]
I would like to give it a shot. 
@jdegoes What was missing from the original @NeQuissimus 's PR?

**Comment by jdegoes** [comment]
@dkarlinsky It was a good start but there are some details I wanted to modify in creating ZIO Test. ZIO Test is a different code base than the original PR but also models tests as effects that produce results.

It's basically doing what Tim did, but for the current ZIO Test code base, and using `RunnableSpec` as the basis.

**Comment by dkarlinsky** [comment]
Ok, I'll try to make it work with the new `RunnableSpec`.

**Comment by jdegoes** [comment]
Awesome, thank you!

The new `RunnableSpec` doesn't actually run yet but you can get the code in place...

**Comment by dkarlinsky** [comment]
Hmm... Is there a way for me to make it run somehow, to test what I'm doing?

**Comment by jdegoes** [comment]
Put a `println("Hello World")` in the main function of `RunnableSpec`?

**Comment by dkarlinsky** [comment]
Sounds pure :)
I think I'm missing a lot of context on how `RunnableSpec` is supposed to work, though...
The main method is final, and calls `System.exit()`, so it seems that the only way to run it from a custom sbt `Framework` is to create a new JVM process and wait for its result code, to figure out if the test succeeded...

**Comment by jdegoes** [comment]
@dkarlinsky Probably we need a separate function, aside from `main`, to call. I will add that function, it will be called `run` and take some arguments.

**Comment by jdegoes** [comment]
@dkarlinsky Do you need success / failure information? What's the ideal information you would obtain from `RunnerSpec.run`?

**Comment by dkarlinsky** [comment]
Well, I see that sbt's `Runner.execute()` method (which I will need to implement) gets an event handler and seems to be expected to report test results as events to the handler.
Here's the [Event interface](https://github.com/sbt/test-interface/blob/master/src/main/java/sbt/testing/Event.java).
Looks like I will need need:  
-  a status (which I will need to map to [Status](https://github.com/sbt/test-interface/blob/master/src/main/java/sbt/testing/Status.java))
-  an `Option[Throwable]`
-  and `duration: Long` (I imagine my runner can measure this)

If there are multiple results that can come from a single `RunnableSpec`, I will also need a label for each one.

**Comment by jdegoes** [comment]
@dkarlinsky 

[This should be enough](https://github.com/zio/zio/pull/1264), will merge when it's green.

**Comment by dkarlinsky** [comment]
Thanks @jdegoes ! 
I started working on it.
I have a some questions:

1.  When I load a`RunnableSpec` object using reflection, I don't see another way but to cast it to `RunnableSpec[Any, String]`. Is this reasonable?
2. I'm having trouble figuring out how to get test results with their corresponding labels from `ExecutedSpec[String]`. I need to get the structure of suite labels with nested test labels and results.
3. It would also be nice to get same tests and suites labels info from the the `Spec` before starting execution, so I can report them as `pending`.

Thanks

**Comment by jdegoes** [comment]
1. Instead, `map` the `Spec`, changing each `L` into a `String` by calling `toString`. To deal with `R`, you can create a wrapper: `SomeRunnerSpec { type R; val spec: RunnableSpec[R, String] }`. Then you can safely / type safely ensure that the runner can deal with the `R`.
2. You can `fold` over the spec, or just pattern match against it recursively. You may want to create an intermediate data structure better aligned for the needs of the integration (but you probably don't need it, strictly speaking).
3. You can do that already, `Spec` can be fully introspected. Each test is an effect, so you don't want to run that in the initial scan, but you can get the structure of the spec and all the labels out pretty easily. After you do (2), it will be more obvious how to do (3).

Let me know if you need help on 2 - 3!

**Comment by dkarlinsky** [comment]
Ok, more questions :)

I'm missing something wrt the wrapper in (1)?  
Say I create a wrapper:
```scala
abstract class SomeRunnerSpec { type R; val spec: RunnableSpec[R, String] }
```
How does this help me to deal with `R`? 
I will have to implement it at some point, and will need to decide on the `R`...

**Comment by jdegoes** [comment]
@dkarlinsky You don't have to deal with `R`, because the spec has a runner in it, that can deal with the r. So what you want is an existential type that can represent the fact that it is some unknown type. 

But actually, in your first version, I think you can just cast `R` as `Any` and ignore it. It won't be precise but I can help you fix it later.r

**Comment by dkarlinsky** [comment]
Ok, sure. 
Thanks!

**Comment by dkarlinsky** [comment]
@jdegoes Another quick question:
I made a `RunnableSpec` based on the example in `ZSpec` scaladoc:
``` scala
object DummySpec extends DefaultRunnableSpec ({
    suite("clock") {
      testM("time is non-zero") {
        assertM(clock.nanoTime.map(_.toInt), lt(0))
      }
    }
})
```
It is asserting that `nanoTime` is negative, so I expected it to fail.
When I run it I however I get exit code 0...
Is this expected behavior?

**Comment by jdegoes** [comment]
@dkarlinsky That is not expected behavior. There are no tests for the testing library yet. 😄

**Comment by dkarlinsky** [comment]
@jdegoes I'm finished with the PR, as far as the functionality, and review comments.
Let me know if you see anything else I need to improve.
Thanks.

**Comment by ghostdogpr** [comment]
Closed by #1335
